user-permissions.md

```diff
@@ -0,0 +1,248 @@
+
+```
+admin_role = Role.new(name: :admin)
+user.roles << admin_role
+
+PermissionChecker.can_view_statements?(user)
+```
+
+**Advantages**
+
+- Better organisation of permissions (centralising the logic)
+- Users can have multiple roles, and roles can have distinct permissions
+- Adding a new role doesn't require a migration
+
+**Disadvantages**
+
+- We can't achieve per-resource authorisation logic (easily)
+- We're on our way to rolling our own authorisation logic when an off-the-self solution may make more sense
+
+### Using an authorisation library
+
+By using an off-the-self authorisation gem we would get the maximum amount of flexibility in terms of authorisation, but at the cost of added complexity and another dependency. It may only make sense to go this route if we expect to need resource-level authorisation logic.
+
+**Advantages**
+
+- Clean modelling or roles and permissions
+- Users can have multiple roles/permissions
+- Users can have resource-level permissions (such as granting `user_a` access to `application_b`)
+- The heavy lifting is done for us (we don't need to roll our own logic)
+
+**Disadvantages**
+
+- Potentially more work up-front
+- May end up needlessly over-complicating things
+
+There are a [number of potential libraries](https://www.ruby-toolbox.com/categories/rails_authorization) that would work well with `devise`, the most common being:
+
+- [Pundit](https://github.com/varvet/pundit)
+  - Policy-based authorisation (define policies for resource/models and rules for access)
+  - Granular control
+  - Explicit authorisation rules defined in policies
+  - A more recent evolution on this gem is [ActionPolicy](https://github.com/palkan/action_policy)
+  - We use this in ECF at the moment
+- [CanCanCan](https://github.com/CanCanCommunity/cancancan)
+  - Ability-based authorisation (for different user roles and contexts)
+  - Clear way of defining rules that govern access to actions and resources
+  - Provides a DSL for clearly defining authorisation rules
+  - A simpler take on the same kind of pattern is [AccessGranted](https://github.com/chaps-io/access-granted)
+- [Rolify](https://github.com/RolifyCommunity/rolify)
+  - Role-based authorisation
+  - Provides convenient methods for scoping based on roles
+  - Simplest to setup and still quite flexible
+  - Not maintained as well as the others
+
+If it was better maintained I would be leaning towards Rolify as a nice balance of simplicity and flexibility - it probably fits our use case the best. That being said, if we feel rolling our own solution doesn't make sense its likely worth going with one of the big two, so Pundit or CanCanCan.
+
+## Final thoughts
+
+- Would we be interested in leveraging an authorisation system for lead providers/managing access in general?
+- Do any of these have a bearing on future integration of DfE sign in?
+
+## Next steps
```

Comment:
not sure why this isn't appearing correctly in the markdown view?

Reply:
bah I get this some times not sure why, very unhelpful diff - must be the kind of space?! 😅 

```
-## Next steps
+## Next steps
```